### ISSUE TITLE
"Install required packages ubuntu/debian" command is missing qttools5-dev package in how_to_build_linux.md

### ISSUE THREAD

**Issue by Koboldthegreat** [ideasman42/opentoonz#11]
Without qttools5-dev package installed, cmake fails to write build files:

    Could not find a package configuration file provided by "Qt5LinguistTools"
    with any of the following names:

    Qt5LinguistToolsConfig.cmake
    qt5linguisttools-config.cmake

**Comment by andreygursky** [comment]
That's not necessary, since a needed package is already mentioned in the doc:
> apt-get install [...] qttools5-dev-tools [...]

which works for me. Moreover qttools5-dev-tools is more "popular" than qttools5-dev: it is contained on the Debian DVD-1 while qttools5-dev is on DVD-4.

**Comment by Koboldthegreat** [comment]
Well, It didn't work on Ubuntu :)

2016-04-16 17:12 GMT+02:00 andreygursky <notifications@github.com>:

**Comment by andreygursky** [comment]
Ubuntu 15.10, 16.04? It would be interesting to know what qt version is in use.

**Comment by andreygursky** [comment]
And cmake version also.

**Comment by Koboldthegreat** [comment]
It's Ubuntu 15.10,
qt: 5.4.2
cmake: 3.4.1

Now the make command fails at this:
 [  0%] Building CXX object
tnzcore/CMakeFiles/tnzcore.dir/__/common/tvrender/tfont_qt.cpp.o
In file included from /usr/include/c++/5/backward/strstream:51:0,
                 from
/home/simon/opentoonz/toonz/sources/include/tcommon.h:27,
                 from
/home/simon/opentoonz/toonz/sources/include/tpixelgr.h:6,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:24:
/usr/include/c++/5/backward/backward_warning.h:32:2: warning: #warning This
file includes at least one deprecated or antiquated header which may be
removed without further notice at a future date. Please use a
non-deprecated interface with equivalent functionality instead. For a
listing of replacement headers and interfaces, consult the file
backward_warning.h. To disable this warning use -Wno-deprecated. [-Wcpp]
 #warning \
  ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp: In member
function ‘void TFontManager::loadFontNames()’:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:392:35:
error: invalid use of incomplete type ‘class QStringList’
  if (m_pimpl->m_qfontdb->families().empty())
                                   ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp: In member
function ‘void TFontManager::setFamily(std::__cxx11::wstring)’:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:406:14:
error: variable ‘QStringList families’ has initializer but incomplete type
  QStringList families = m_pimpl->m_qfontdb->families();
              ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:406:54:
error: invalid use of incomplete type ‘class QStringList’
  QStringList families = m_pimpl->m_qfontdb->families();
                                                      ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp: In member
function ‘void TFontManager::setTypeface(std::__cxx11::wstring)’:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:432:14:
error: variable ‘QStringList styles’ has initializer but incomplete type
  QStringList styles =
m_pimpl->m_qfontdb->styles(QString::fromStdWString(m_pimp
              ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:432:99:
error: invalid use of incomplete type ‘class QStringList’

m_pimpl->m_qfontdb->styles(QString::fromStdWString(m_pimpl->m_currentFamily));

  ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
In file included from
/home/simon/opentoonz/toonz/sources/include/tcommon.h:23:0,
                 from
/home/simon/opentoonz/toonz/sources/include/tpixelgr.h:6,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:24:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp: In member
function ‘TFont* TFontManager::getCurrentFont()’:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:479:39:
error: invalid use of incomplete type ‘class QStringList’
  assert(!m_pimpl->m_qfontdb->families().empty());
                                       ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:480:41:
error: invalid use of incomplete type ‘class QStringList’
  setFamily(m_pimpl->m_qfontdb->families().first().toStdWString());
                                         ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp: In member
function ‘void
TFontManager::getAllFamilies(std::vector<std::__cxx11::basic_string<wchar_t>
>&) const’:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:489:14:
error: variable ‘QStringList qFamilies’ has initializer but incomplete type
  QStringList qFamilies = m_pimpl->m_qfontdb->families();
              ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:489:55:
error: invalid use of incomplete type ‘class QStringList’
  QStringList qFamilies = m_pimpl->m_qfontdb->families();
                                                       ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:494:2:
error: incomplete type ‘QStringList’ used in nested name specifier
  QStringList::const_iterator it = qFamilies.begin();
  ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:495:9:
error: ‘it’ was not declared in this scope
  for (; it != qFamilies.end(); ++it) {
         ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp: In member
function ‘void
TFontManager::getAllTypefaces(std::vector<std::__cxx11::basic_string<wchar_t>
>&) const’:
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:506:14:
error: variable ‘QStringList qStyles’ has initializer but incomplete type
  QStringList qStyles =
m_pimpl->m_qfontdb->styles(QString::fromStdWString(m_pim
              ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:506:100:
error: invalid use of incomplete type ‘class QStringList’

m_pimpl->m_qfontdb->styles(QString::fromStdWString(m_pimpl->m_currentFamily));

  ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtGui/qfont.h:38:0,
                 from /usr/include/x86_64-linux-gnu/qt5/QtGui/QFont:1,
                 from
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:12:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:73:7: note: forward
declaration of ‘class QStringList’
 class QStringList;
       ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:512:2:
error: incomplete type ‘QStringList’ used in nested name specifier
  QStringList::const_iterator it_typeface = qStyles.begin();
  ^
/home/simon/opentoonz/toonz/sources/common/tvrender/tfont_qt.cpp:513:9:
error: ‘it_typeface’ was not declared in this scope
  for (; it_typeface != qStyles.end(); ++it_typeface) {
         ^
tnzcore/CMakeFiles/tnzcore.dir/build.make:2970: recipe for target
'tnzcore/CMakeFiles/tnzcore.dir/__/common/tvrender/tfont_qt.cpp.o' failed
make[2]: ***
[tnzcore/CMakeFiles/tnzcore.dir/__/common/tvrender/tfont_qt.cpp.o] Error 1
CMakeFiles/Makefile2:86: recipe for target
'tnzcore/CMakeFiles/tnzcore.dir/all' failed
make[1]: *** [tnzcore/CMakeFiles/tnzcore.dir/all] Error 2
Makefile:75: recipe for target 'all' failed
make: *** [all] Error 2
simon@UbuntuDesktop:~/opentoonz/toonz/sources$ cmake --version
cmake version 3.4.1

CMake suite maintained and supported by Kitware (kitware.com/cmake).




2016-04-16 18:38 GMT+02:00 andreygursky <notifications@github.com>:

**Comment by andreygursky** [comment]
I believe there has been already reported a similar issue. While it is likely that OpenToonz could be built with Qt4, for Qt5 some features from 5.5 are needed.

**Comment by andreygursky** [comment]
However it might be something trivial like:
```diff
diff --git a/toonz/sources/common/tvrender/tfont_qt.cpp b/toonz/sources/common/tvrender/tfont_qt.cpp
index accb227..66ec8f7 100644
--- a/toonz/sources/common/tvrender/tfont_qt.cpp
+++ b/toonz/sources/common/tvrender/tfont_qt.cpp
@@ -9,6 +9,7 @@
 //#include FT_FREETYPE_H
 
 
+#include <QStringList>
 #include <QFont>
 #include <QFontDatabase>
 #include <QFontMetrics>
```
Apply this patch like:
```
cd /home/simon/opentoonz
patch -p1 < [filename where you saved the patch]
```
then continue with building (cd in your top build dir). And let us know the result.

**Comment by Koboldthegreat** [comment]
Installing qt5.5 via this ppa:
https://launchpad.net/~canonical-qt5-edgers/+archive/ubuntu/backup-qt551
fixed everything!

thx for everything!

2016-04-16 19:31 GMT+02:00 andreygursky <notifications@github.com>:

**Comment by Koboldthegreat** [comment]
https://gyazo.com/4526aec081455d9cc70182cd10af9293

2016-04-16 20:56 GMT+02:00 Simon Bil <simontje.bil@gmail.com>:

**Comment by mmuman** [comment]
Well, I wrote this file using Qt5.5 here, so if adding the include fixes the build with Qt5.4 it shouldn't harm adding it.

As for the packages, I'm on Debian Sid here, and apt-cache says:
```
# apt-cache show qttools5-dev-tools 
Package: qttools5-dev-tools
…
Replaces: qttools5-dev (<< 5.5.0-2~), qttools5-examples (<< 5.3.0-4~)
Breaks: qttools5-dev (<< 5.5.0-2~), qttools5-examples (<< 5.3.0-4~)
```
Which means older versions likely require qttools5-dev instead. We can probably mention that in the docs.

**Comment by andreygursky** [comment]
Yes, I've compiled a minimal example on 4.8.7. Including QFont, QString is not enough. One should include QStringList explicitly. I've commited this.
> Which means older versions likely require qttools5-dev instead. We can probably mention that in the docs.

Yes. Sadly it will break a one universal apt-get line (though we can add unneeded but harmless qttools5-dev also) or add few more notes.

**Comment by mmuman** [comment]
I'd have used a squash commit for this, so the next rebase would have cleaned it up...

**Comment by andreygursky** [comment]
Wouldn't squashing affect how the things are evolving? I find preserving all big and small steps and credits encourage contributions. As long as they don't get lost, no problem. During further testing other caveats will come up making an ideal squashed commit unlikely (and not authentic). @mmuman, or you mean something else?

**Comment by mmuman** [comment]
Yes but keeping commits for oneliners doesn't help readability, furthermore I wanted to squash the references to FreeType in this file as I originally intended on using it until I figured out I could just get along with only Qt.

**Comment by ideasman42** [comment]
Suggest we continue to extract PR's and submit them upstream, recently PR's have been getting accepted more quickly, so we might not need to keep this branch around for so long.
